mlx/backend/common/quantized.cpp

```diff
@@ -151,6 +151,78 @@ void _qmm_t(
   }
 }
 
+template <int bits, int S>
+simd::Simd<uint32_t, S> extract_bits_simd(const uint32_t* w) {
+  constexpr int bitmask = (1 << bits) - 1;
+  simd::Simd<uint32_t, S> wi;
+  if constexpr (bits == 4 && S == 8) {
+    constexpr std::array<uint32_t, 8> shifts_ = {{0, 4, 8, 12, 16, 20, 24, 28}};
+    auto shifts(*(simd::Simd<uint32_t, S>*)&shifts_);
+    wi = simd::Simd<uint32_t, S>(*w);
+    wi = wi >> shifts;
+    wi = wi & bitmask;
+  } else if constexpr (bits == 8 && S == 8) {
+    constexpr std::array<uint32_t, 8> shifts_ = {{0, 8, 16, 24, 0, 8, 16, 24}};
+    auto shifts(*(simd::Simd<uint32_t, S>*)&shifts_);
+    auto l = simd::Simd<uint32_t, 4>(*w++);
+    auto r = simd::Simd<uint32_t, 4>(*w);
```

Comment:
Sorry for posting post-merge, but the following is easier to describe in the context of this PR:

Clang on the latest FreeBSD, 13.4, and 14.2, is unhappy with the `sims::Simd` calls in L167-L168: `error: implicit instantiation of undefined template 'mlx::core::simd::Simd<unsigned int, 4>`, cf. 
https://buildkite.com/julialang/yggdrasil/builds/17042#0194c6a1-5b8d-4a7c-8cef-8e098d29b750/6-32274

Not sure how to resolve this...

Reply:
Might be good to file an issue. That way we won't forget about this and we can help debug it.

Reply:
https://github.com/ml-explore/mlx/issues/1823